packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -3677,77 +3692,71 @@ namespace Tpetra {
+
+    const std::pair<size_t, size_t> rowRange (0, numRows);
+    const std::pair<size_t, size_t> colRange (0, numCols);
 
-    // FIXME (mfh 22 Jul 2014, 10 Dec 2014) Currently, it doesn't work
-    // to do a 2-D copy, even if this MultiVector has constant stride.
-    // This is because Kokkos can't currently tell the difference
-    // between padding (which permits a single deep_copy for the whole
-    // 2-D View) and stride > numRows (which does NOT permit a single
-    // deep_copy for the whole 2-D View).  Carter is working on this,
-    // but for now, the temporary fix is to copy one column at a time.
+    input_view_type A_view_orig (reinterpret_cast<IST*> (A.getRawPtr ()),
+				 LDA, numCols);
+    auto A_view = Kokkos::subview (A_view_orig, rowRange, colRange);
 
     // Use the most recently updated version of this MultiVector's
     // data.  This avoids sync'ing, which could violate users'
     // expectations.
     //
     // If we need sync to device, then host has the most recent version.
-    const bool useHostVersion = this->template need_sync<device_type> ();
+    const bool useHostVersion = this->need_sync_device ();
 
     dev_view_type srcView_dev;
     host_view_type srcView_host;
     if (useHostVersion) {
-      srcView_host = getDualView().view_host ();
+      srcView_host = this->getLocalViewHost ();
     }
     else {
-      srcView_dev = this->template getLocalView<device_type> ();
+      srcView_dev = this->getLocalViewDevice ();
     }
 
-    for (size_t j = 0; j < numCols; ++j) {
-      const size_t srcCol =
-        this->isConstantStride () ? j : this->whichVectors_[j];
-      const size_t dstCol = j;
-      IST* const dstColRaw =
-        reinterpret_cast<IST*> (A.getRawPtr () + LDA * dstCol);
-      input_col_type dstColView (dstColRaw, numRows);
-
+    if (this->isConstantStride ()) {
       if (useHostVersion) {
-        auto srcColView_host = Kokkos::subview (srcView_host, rowRange, srcCol);
-        TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
-          dstColView.extent (0) != srcColView_host.extent (0),
-          std::logic_error, ": srcColView and dstColView_host have different "
-          "dimensions.  Please report this bug to the Tpetra developers.");
-        Kokkos::deep_copy (dstColView, srcColView_host);
+	Kokkos::deep_copy (A_view, srcView_host);
       }
       else {
-        auto srcColView_dev = Kokkos::subview (srcView_dev, rowRange, srcCol);
-        TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(
-          dstColView.extent (0) != srcColView_dev.extent (0),
-          std::logic_error, ": srcColView and dstColView_dev have different "
-          "dimensions.  Please report this bug to the Tpetra developers.");
-        Kokkos::deep_copy (dstColView, srcColView_dev);
+	Kokkos::deep_copy (A_view, srcView_dev);
+      }
+    }
+    else {
+      for (size_t j = 0; j < numCols; ++j) {
+	const size_t srcCol = this->whichVectors_[j];
+	auto dstColView = Kokkos::subview (A_view, rowRange, j);
```

Comment:
@trilinos/framework @GeoffDanielson @william76 @bartlettroscoe @jwillenbring @prwolfe  You can literally see that I've removed the line of code that shows up in PR testing as an unused variable warning.